### PR TITLE
Fees module refactor

### LINKED ISSUE #1993: Fees module questions
Follow up of https://github.com/paritytech/substrate/pull/1978#issuecomment-472327208 https://github.com/paritytech/substrate/pull/1815#discussion_r265284213 https://github.com/paritytech/substrate/pull/1924#issuecomment-472415736

It is clear to me that my understanding of fees model is different to the one @gavofyork has in mind.
I would like to have this clearly documented to make sure everyone including future developers understands the fee model of substrate.

The questions are: 
- What is `TransactionPayment` and the purpose of it.
- What is transaction fee and the relationship between transaction fee and `TransactionPayment`.
  - My understanding is that transaction fee is all the fees user pays to have a transaction confirmed.
  - For balances.transfer, it is `base fee + tx size * byte fee + transfer/creation fee`
  - For contract.call, it is `base fee + tx size * byte fee + gas fee`

To my understanding, the purpose of transaction fee is to ensure the sender pays for at least of cost for the network to process this transaction. It also acts as a security gateway to ensure the cost to spam the network is high.

Therefore, the transaction fee should be direct proportion to the resources (CPU, storage, network) required to process this transaction. As a result, in addition to the mandatory fee `base fee + tx size * byte fee` that charged by executive module, each other runtime call should have additional fee that reflects on the complexity of the call. For example, create a new account with balances.transfer requires additional storage usage, and hence a higher account creation fee.

Without this feature, attacker could identify an expensive runtime call and use it to attack the network efficiently. Unless, of cause, all the runtime calls are simple and have O(1) computation complexity and storage usage, which I don't think will be the case for most of the parachains.

In my design, the fees module should achieve those goals:
- Centralized place to query fees charged for each transaction
  - This is achieved by `fees.Charged` event. A single event will be emitted for each transaction and clients (apps) can easily query this information.
- Centralized place to charge transaction fee
  - Instead of have each module use `Currency` or `TransferAsset` trait to modify the balance and charge fee, `ChargeFee` trait offers a unified way to charge fee. Fees module implementation will handle all the balance checking, liquidity checking, and accumulate the total fee.
  - Otherwise it is easy to have buggy code. #1980 
- Enable the possibility to query fees information from other runtime modules
  - This is achieved by `OnFeeCharged` trait #1924 

-----

Next steps:

I can see those options to proceed forward:

- Close #1815 and #1924 and update existing fees module to avoid storage usage. Ensure the abstraction allows developer have those additional logic implemented in their own runtime.
  - This means I can create my own implementation of `ChargeFee` trait and use it instead of the `srml-fees` module
- Or create a in memory temporary store for fees calculation to avoid the trie db overhead.
- Or revert and remove fees module and related changes.

### PR DESCRIPTION
**Pull request by xlc**
Closes #1993

### CODE REVIEW

File: srml/fees/src/lib.rs [hunk 1]
```diff
@@ -44,74 +44,54 @@ pub trait Trait: system::Trait {
...
 impl<T: Trait> ChargeFee<T::AccountId> for Module<T> {
 	type Amount = AssetOf<T>;
 
 	fn charge_fee(transactor: &T::AccountId, amount: AssetOf<T>) -> Result {
-		let extrinsic_index = <system::Module<T>>::extrinsic_index().ok_or_else(|| "no extrinsic index found")?;
-		let current_fee = Self::current_transaction_fee(extrinsic_index);
-		let new_fee = current_fee.checked_add(&amount).ok_or_else(|| "fee got overflow after charge")?;
-
-		T::TransferAsset::withdraw(transactor, amount, WithdrawReason::TransactionPayment)?;
+		T::TransferAsset::withdraw(transactor, amount, WithdrawReason::Reserve)?;
```
Comment (gavofyork) [inline srml/fees/src/lib.rs]: why is this reserve?

**Comment by gavofyork** [comment]
I think the `ChargeFee` module confuses things in general as it presupposes that Substrate actually has a cohesive and uniform notion of `fee`. As I mentioned in the issue, it doesn't - economic protections against spam are implemented in numerous ways, only one of which is a straight fee. Reserve and refund doesn't work well beyond the simple assumptions made here (in particular, what about multiple blocks).

All in all, I would propose to revert the fees module entirely (it was merged without my consent in the first place) and instead introduce a number of ZCAs in any existing modules that do things that you'd want to catch and handle differently. There can then be a fees module introduced which can be hooked into all of these events without compromising the performance or API hygiene of the main SRML modules.

**Comment by xlc** [comment]
@gavofyork I have removed the `ChargeFee` trait and only keep the make transaction payment functionality in fees module.
I still want to keep fees module for `TransactionBaseFee` and `TransactionByteFee`. Previously they belongs to balances module which I don't think it is the most sensible place.

**Comment by gavofyork** [comment]
#2028 is the way things need to go.

This removes all the fees stuff, replacing `ChargeFee` with `MakePayment`. I generally agree that it's not a great fit to have the tx byte/base fee in balances, but I'm also not convinced that setting up a whole other module for it it sensible either. With all the abstractions I introduce in #2028, you should still be able to do everything you need, including a fairly holistic fees amalgamation module.

**Pull request closed without merging**